### PR TITLE
Add support for fchmodat2 and use it where appropriate

### PR DESCRIPTION
**Pull request by YHNdnzj**
This is a refreshed version of #29664, with my own comments resolved and an addition that IMO doesn't provide much value dropped.

### CODE REVIEW

File: src/test/test-seccomp.c [hunk 1]
```diff
@@ -1007,21 +1007,22 @@ static int real_open(const char *path, int flags, mode_t mode) {
...
+
         /* glibc does not provide a direct wrapper for fchmodat2(). Let's hence define our own wrapper for
          * testing purposes that calls the real syscall, on architectures and in environments where
          * SYS_fchmodat2 is defined. Otherwise, let's just fall back to the glibc fchmodat() call. */
 
-#if defined __NR_fchmodat2 && __NR_fchmodat2 >= 0
-        int r;
-        r = (int) syscall(__NR_fchmodat2, dirfd, path, flags, mode);
+        /* Not supported by fchmodat() */
+        assert_se(!FLAGS_SET(flags, AT_EMPTY_PATH));
```
Comment (poettering) [inline src/test/test-seccomp.c:1018]: maybe we should just make it work ourselves? i if the flag is set and path isempty(path) holds, then redirect to fchmod()?

File: src/basic/fs-util.c [hunk 2]
```diff
@@ -288,8 +288,22 @@ int fchmod_umask(int fd, mode_t m) {
...
+         * - fchmodat2(2) supports all the AT_* flags, but is still very recent.
+         *
+         * We try to use fchmodat2(), and, if it is not supported, resort
+         * to the /proc/self/fd dance. */
+
+        assert(fd >= 0);
+
+        if (fchmodat2(fd, "", m, AT_EMPTY_PATH) >= 0)
+                return 0;
+        if (errno != ENOSYS)
```
Comment (poettering) [inline src/basic/fs-util.c]: hmm, we should also accept EPERM here, some container mgrs block unknown syscalls with EPERM

File: src/nspawn/nspawn-patch-uid.c [hunk 3]
```diff
@@ -239,14 +240,18 @@ static int patch_fd(int fd, const char *name, const struct stat *st, uid_t shift
...
                 if (name) {
+                        /* It looks like older glibc (before 2016) did not support AT_SYMLINK_NOFOLLOW. */
                         if (!S_ISLNK(st->st_mode))
-                                r = fchmodat(fd, name, st->st_mode, 0);
-                        else /* AT_SYMLINK_NOFOLLOW is not available for fchmodat() */
-                                r = 0;
+                                r = RET_NERRNO(fchmodat(fd, name, st->st_mode, 0));
+                        else {
+                                r = RET_NERRNO(fchmodat2(fd, name, st->st_mode, AT_SYMLINK_NOFOLLOW));
+                                if (r == -ENOSYS)
```
Comment (poettering) [inline src/nspawn/nspawn-patch-uid.c]: as above

**Comment by poettering** [comment]
lgtm

**Review by yuwata — APPROVED** [review]

**Pull request merged**